Docs: Add initial glossary terms (#13465)
gatsbyjs/gatsby: docs/docs/glossary.md

Allows you to customise the behavior of your app depending on its [environment](#environment).

## F

### Filesystem

The way files are organized. But with Gatsby it means having files in the same place as your website's or app's code instead of having data from an external [source](#source)

### Front-End

The [public-facing](#public) interface for your website or app.

## G

### Gatsby

This project. A framework based on [React](#react) that helps you build blazing fast websites and apps.

### GraphQL

A [query](#query) language that allows you to pull data into your website or app. It’s the interface that Gatsby uses for managing site data.

## H

### Hosting

A hosting provider keeps a copy of your website or app and makes it accessible to [the public](#public).

### HTML

A markup language that every web browser is able to understand. It stands for HyperText Markup Language. HTML gives your web content a universal informational structure, defining things like headings, paragraphs, and more.

### Headless CMS

A [CMS](#cms) that only handles the [back-end](#back-end) content management instead of handling both the back-end and [front-end](#frontend).

## I

> ## J

Reviewer: I thought of one more for J: `JAMstack`!
```
JAMStack refers to a modern web architecture using [JavaScript](#javascript), [APIs](#api), and ([HTML](#html)) markup. From [JAMStack.org](https://jamstack.org): "It’s a new way of building websites and apps that delivers better performance, higher security, lower cost of scaling, and a better developer experience."
```

### NPM

[Node](#node) Package Manager. Allows you to install and update other packages that your project depends on. [Gatsby](#gatsby) and [React](#react) are examples of your project's dependencies. See also: [Yarn](#yarn).

## O

## P

### Page

An [HTML](#html) page.

This also often refers to [components](#components) that live in `/src/pages/` and are converted to pages by Gatsby.

### Plugin

Additional code that adds functionality to Gatsby that wasn't included out-of-the-box.

### Public

This usually refers to either a member of the public (as opposed to your team) or the folder `/public` in which your [built](#build) website or app is saved.

### Programmatically

Something that automatically happens based on your code and configuration. For example, you might [configure](#config) your project to create a [page](#page) for every blog post written.

### Production Environment

The [environment](#environment) for the [built](#build) website or app. It's accessed through the [CLI](#cli) using `gatsby build` or `gatsby serve`.

## Q

### Query

The process of requesting specific data from somewhere. With Gatsby you normally query with [GraphQL](#graphql).

> ## R

Reviewer: We could add "Router" here to talk a bit about the client-rendering part of Gatsby.  https://www.gatsbyjs.org/docs/routing/

Author: Could you give me a hand with this definition please? Struggling to come up with something that doesn't sound strange and I'm not sure my definition is actually correct.

```
Routing allows [Pages](#pages) and [Components](#components) to be accessible via a URL.
```

Reviewer: Sure thing! I wrote about routing in a [blog post](https://marcysutton.com/links-vs-buttons-in-modern-web-applications) once...how about something like this?
> Much like filtering in a mail room, routing is the mechanism for connecting network requests with the appropriate content in a web application. Routers like `@reach/router`–used in Gatsby–are common in JavaScript web frameworks: they map URL fragments with views and subviews, enabling dynamic URLs that don't have to be written by hand.

Author: Cool, thanks. I submitted a tweaked version but strangely I had to do a new pull request for it (I'm probably misusing GitHub) https://github.com/gatsbyjs/gatsby/pull/13739

Reviewer: It wasn't you, there was a configuration change that needed some adjustment. It should be resolved soon if it hasn't already!

### React

> A code library (built with [JavaScript](#javascript)) for building user interfaces. It’s the framework that [Gatsby](#gatsby) uses to build pages and structure content.

Reviewer: ```suggestion
A code library (written with [JavaScript](#javascript)) for building user interfaces. It’s the framework that [Gatsby](#gatsby) uses to build pages and structure content.
```
Diversifying word choice since it says "building" again in regards to user interfaces.

## S

### Source Code

> Source code is your code that lives in `/src/` folder and makes up the unique aspects of your website or app. It is made up of [JavaScript](#javascript) and sometimes [CSS](#css) and other files.

Reviewer: Would this include Markdown files? Or should we add a Content section for those?

Author: I've note really used Markdown on Gatsby so I'm unsure on where the conventional place is to keep them. What would you recommend?

Reviewer: On second thought, I think it's fine as-is. 👍

The source code gets [built](#build) into the site the [public](#public) will see.

### Source Plugin

> A [plugin](#plugin) that adds additional [data sources](#data-source) to Gatsby that can then be [queried](#query) by your [pages](#page) and [components](#component).

Reviewer: Nice! 👍

### Starter

> A pre-configured Gatsby project that can be used as a starting point for your project. They can be discovered using the [Gatsby Starter Library](/starters/).

Reviewer: ```suggestion
A pre-configured Gatsby project that can be used as a starting point for your project. They can be discovered using the [Gatsby Starter Library](/starters/) and installed using the [Gatsby CLI](/docs/starters/).
```

### Static

Gatsby [builds](#build) static versions of your page that can be easily [hosted](#hosting). This is in contrast to dynamic systems in which each page is generated on-the-fly. Being static affords major performance gains because the work only needs to be done once per content or code change.

It also refers to the `/static` folder which is automatically copied into `/public` on each [build](#build) for files that don't need to be processed by Gatsby but do need to exist in [public](#public).

## T

### Template

A [component](#component) that is [programatically](#programatically) turned into a page by Gatsby.

### Transformer

A [plugin](#plugin) that transforms one type of data to another. For example you might transform a spreadsheet into a [JavaScript](#javascript) array.

## U

## V

## W

### Webpack

A [JavaScript](#javascript) application that Gatsby uses to bundle your website's code up. This happens automatically on [build](#build).

## X

## Y

### Yarn

> A package manager that some prefer to [NPM](#npm).

Reviewer: ```suggestion
A package manager that some prefer to [NPM](#npm). It is also required for [developing Gatsby](/contributing/setting-up-your-local-dev-environment/#using-yarn).
```

## Z

> ## 0-9

Reviewer: I don't think we need the numbers here, but the letter placeholders are nice!

### CMS

Content Management System: an application where you can manage your content and have it saved to a database or file for accessing later. Examples of Content Management Systems include Wordpress, Drupal, Contentful, and Netlify CMS.

### Command Line

A text-based interface to run commands on your computer. The default Command Line applications for Mac and Windows are `Terminal` and `Command Prompt` respectively.

### Compiler

A compiler is a program that translates code written in one language to another language. For example [Gatsby](#gatsby) can compile [React](#react) applications into static [HTML](#html) files.

### Components

Components are independent and re-usable chunks of code powered by [React](#react) that, when combined, make up your website or app.

A component can include components within it. In fact, [pages](#page) and [templates](#templates) are examples of components.

### Config

The configuration file, `gatsby-config.js` tells Gatsby information about your website. A common option set in config is your sites meta data that can power your SEO meta tags.

## D

### Data Source

> A source of data that usually gets fed into Gatsby using [source plugins](#source-plugin). A data source is often a [Headless CMS](#headless-cms), but it could also include Markdown, JSON, or YAML files.

Reviewer: ```suggestion
Content and data's origin point, typically integrated into Gatsby with [source plugins](#source-plugin). A data source is often a [Headless CMS](#headless-cms), but it could also include Markdown, JSON, or YAML files.
```

### Database

A database is a structured collection of data or content. Often a [CMS](#cms) will save to a database using [back-end technologies](#back-end). They're often accessed in Gatsby via a [source plugin](#source-plugin)

### Decoupled

> Decoupling describes the separation of different concerns. With [Gatsby](#gatsby) this most commonly means decoupling the [Front-End](#front-end) from the [Back-End](#back-end)

Reviewer: ```suggestion
Decoupling describes the separation of different concerns. With [Gatsby](#gatsby) this most commonly means decoupling the [Front-End](#front-end) from the [Back-End](#back-end), like with [Decoupled Drupal](https://dri.es/how-to-decouple-drupal-in-2019) or [Headless Wordpress](https://www.smashingmagazine.com/2018/10/headless-wordpress-decoupled/).
```

## E

### ECMAScript

ECMAScript (often referred to as ES) is a specification for scripting languages. [JavaScript](#javascript) is an implementation of ECMAScript. Often developers will use [Babel](#babel) to [compile](#compiler) the latest ECMAScript code into more widely supported JavaScript.

### Environment

The environment that Gatsby runs in. For example, when you are writing your code you probably want as much debugging as possible, but that's undesirable on the live website or app. As such, Gatsby can change its behavior depending on the environment it's in.

Gatsby supports two environments by default, the [development environment](#development-environment) and the [production environment](#production-environment).

### Environment Variables

[Environment Variables](/docs/environment-variables/) allow you to customise the behavior of your app depending on its [environment](#environment). For instance, you may wish to get content from a staging CMS during development and connect to your production CMS when you [build](#build) your site. With environment variables you can set a different URL for each environment.

## F

### Filesystem

> The way files are organized. With Gatsby, it means having files in the same place as your website's or app's code instead of pulling data from an external [source](#source)

Reviewer: ```suggestion
The way files are organized. With Gatsby, it means having files in the same place as your website's or app's code instead of pulling data from an external [source](#source). Common filesystem usage in Gatsby includes Markdown content, images, data files, and other assets.
```

### Front-End

The [public-facing](#public) interface for your website or app, delivered using web technologies: HTML, CSS, and JavaScript.

## G

### Gatsby

> This project. A framework based on [React](#react) that helps you build blazing fast websites and apps. Its [feature set](/features/) can make it a good choice over other ways to build websites.

Reviewer: ```suggestion
Gatsby is a modern website framework that builds performance into every website or app by leveraging the latest web technologies such as [React](#react), [GraphQL](#graphql), and modern [Javascript](#javascript). Gatsby makes it easy to create blazing fast, compelling web experiences without needing to become a performance expert.
```

### Hydration

Once a page that has been generated by Gatsby is loaded in a web browser, it is re-hydrated into a full React application that can manipulate the DOM.

## I

## J

### JAMStack

JAMStack refers to a modern web architecture using [JavaScript](#javascript), [APIs](#api), and ([HTML](#html)) markup. From [JAMStack.org](https://jamstack.org): "It’s a new way of building websites and apps that delivers better performance, higher security, lower cost of scaling, and a better developer experience."

### JavaScript

A programming language that helps us make the web dynamic and interactive. [JavaScript](https://developer.mozilla.org/en-US/docs/Web/Javascript) is a widely deployed web technology in browsers. It is also used on the server-side with [Node.js](#node). It is an implementation of the [ECMAScript](#ECMAScript) specification.

## K

## L

## M

### MDX

Extends [Markdown](#markdown) to support [React](#react) [components](#components) within your content.

### Markdown

> A way of writing content in plain text, using special characters such as underscores and asterisks to format content.

Reviewer: ```suggestion
A way of writing HTML content with plain text, using special characters to denote content types such as hash symbols for [headings](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements), and underscores and asterisks for text emphasis.
```
